### DISCUSSION TITLE
WASM SmartWeave contracts

### DISCUSSION THREAD

**Discussion by jsoneaday** [warp-contracts/warp#49] [Ideas]
So smartweave contracts run on the user client correct? I was wondering if you had any plans to allow other languages like Swift?

**Comment by ppedziwiatr** [comment]
> So smartweave contracts run on the user client correct? I was wondering if you had any plans to allow other languages like Swift?

Hey, we're planning to implement https://github.com/redstone-finance/redstone-smartcontracts/issues/27 - if we decide to follow the WASM path, you'll be able to write your contracts in any WASM-compatible language - C/C++, Go, AssemblyScript, Kotlin - I believe that Swift is also supported.

**Comment by jsoneaday** [comment]
This is super interesting. I am the developer for DeChat a social media app on Arweave. I am writing the front end in Swift so this would definitely be needed for me.
Also as you showed if done in Wasm you need not worry about having to support multiple languages since so many languages, including Swift, emit to Webassembly.
Any eta on this rolling out? By way C++, Rust and Swift work with Wasm right now.

**Reply by ppedziwiatr** [reply]
Re. ETA...it will be probably taken into development next week...but it's really hard to estimate, as it will require a lot of R&D. Trying to be realistic here - I think PoC should be presented by the end of this month.

We're now implementing another very complicated (and - as we believe - required by real-life applications) feature - writes between contracts - it will be released soon as an experimental feature (available after switching a dedicated feature switch).

**Comment by jsoneaday** [comment]
Looking forward to all your work. Including the experimental write between contracts feature. Thanks for doing this for the community!

**Reply by ppedziwiatr** [reply]
Re. writes between contracts - https://github.com/redstone-finance/redstone-smartcontracts/discussions/50
Re. WASM stuff - we will keep posting updates in this thread

**Comment by ppedziwiatr** [comment]
as a small update - we're currently researching a contract written in AssemblyScript - all work is being done in this repo https://github.com/redstone-finance/redstone-contracts-wasm

We're currently trying to develop the contract as a class (with a well defined interface) and we're simulating interactions with the contract from the host (ie. accessing contract class, calling functions with arguments type other than numeric (which turns out to be no that obvious in WASM) and also dumping/loading WASM module memory - as a simulation of writing/reading contract state to/from cache).

The contract code is here: https://github.com/redstone-finance/redstone-contracts-wasm/blob/main/assembly/RedStoneToken.ts
Host code is here: https://github.com/redstone-finance/redstone-contracts-wasm/blob/main/class.js
